sui/src/benchmark.rs

```diff
@@ -0,0 +1,289 @@
+            (host, port, type_)
+        } else {
+            panic!("Invalid variant")
+        };
+
+    let network_client = NetworkClient::new(
+        host.clone(),
+        port,
+        benchmark.buffer_size,
+        Duration::from_micros(benchmark.send_timeout_us),
+        Duration::from_micros(benchmark.recv_timeout_us),
+    );
+    let network_server = NetworkServer::new(host, port, benchmark.buffer_size);
+    let connections = if benchmark.tcp_connections > 0 {
+        benchmark.tcp_connections
+    } else {
+        num_cpus::get()
+    };
+
+    match type_ {
+        MicroBenchmarkType::Throughput { num_transactions } => run_throughout_microbench(
+            network_client,
+            network_server,
+            connections,
+            benchmark.batch_size,
+            benchmark.use_move,
+            num_transactions,
+            benchmark.committee_size,
+            benchmark.db_cpus,
+        ),
+        MicroBenchmarkType::Latency {
+            num_chunks,
+            chunk_size,
+            period_us,
+        } => run_latency_microbench(
+            network_client,
+            network_server,
+            connections,
+            benchmark.use_move,
+            benchmark.committee_size,
+            benchmark.db_cpus,
+            num_chunks,
+            chunk_size,
+            period_us,
+        ),
+    }
+}
+
+fn run_throughout_microbench(
+    network_client: NetworkClient,
+    network_server: NetworkServer,
+    connections: usize,
+    batch_size: usize,
+    use_move: bool,
+    num_transactions: usize,
+    committee_size: usize,
+    db_cpus: usize,
+) -> MicroBenchmarkResult {
+    assert_eq!(
+        num_transactions % batch_size,
+        0,
+        "num_transactions must integer divide batch_size",
+    );
+
+    assert!(
+        (num_transactions % connections) == 0,
+        "num_transactions must {} be multiple of number of TCP connections {}",
+        num_transactions,
+        connections
+    );
+    let mut tx_cr = TransactionCreator::new(committee_size, db_cpus);
+
+    let chunk_size = batch_size * connections;
+    let txes = tx_cr.generate_transactions(
+        connections,
+        use_move,
+        batch_size * connections,
+        num_transactions / chunk_size,
+    );
+
+    // Make multi-threaded runtime for the authority
+    thread::spawn(move || {
+        let runtime = Builder::new_multi_thread()
+            .enable_all()
+            .thread_stack_size(32 * 1024 * 1024)
+            .build()
+            .unwrap();
+
+        runtime.block_on(async move {
+            let server = spawn_authority_server(network_server, tx_cr.authority_state).await;
+            if let Err(e) = server.join().await {
+                error!("Server ended with an error: {e}");
+            }
+        });
+    });
+
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_stack_size(32 * 1024 * 1024)
```

Comment:
Are these params always the same as the ones in line 122? If so should they be in common?